### PR TITLE
Fix wrong response for Get Snapshots API

### LINKED ISSUE #68090: Return snapshots in progress when ignore_unavailable is set to true in Get Snapshots API
**Elasticsearch version** (`bin/elasticsearch --version`): master,7.5.1,6.8.2

**Plugins installed**: []

**JVM version** (`java -version`): 14

**OS version** (`uname -a` if on a Unix-like system): macos 10.13.6 

**Description of the problem including expected versus actual behavior**:

When executing Get Snapshots API, if the target snapshot does not exist, and `ignore_unavailable` is set to `true`,  then the API returns all the snapshots which are in progress.

**Steps to reproduce**:

 1. Create a snapshot which will take a long time to execute.
 ```
 PUT _snapshot/my_repository/snap1
{
  "indices":["y"]
}
``` 
 2. Get an unknown snapshot, and set `ignore_unavailable` to true
 ```
GET _snapshot/my_repository/xxx?ignore_unavailable=true
```
 then you can see the `snap1` existed in the response.

### PR DESCRIPTION
**Pull request by gaobinlong**
Closes #68090.

The main changes are:
1. fix the bug for Get Snapshots API.
2. add some test code.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/get/TransportGetSnapshotsAction.java [hunk 1]
```diff
@@ -236,8 +236,10 @@ private void loadSnapshotInfos(SnapshotsInProgress snapshotsInProgress, String r
         final List<SnapshotsInProgress.Entry> entries = SnapshotsService.currentSnapshots(
             snapshotsInProgress, repositoryName, snapshotIdsToIterate.stream().map(SnapshotId::getName).collect(Collectors.toList()));
         for (SnapshotsInProgress.Entry entry : entries) {
-            snapshotSet.add(new SnapshotInfo(entry));
-            snapshotIdsToIterate.remove(entry.snapshot().getSnapshotId());
+            if (snapshotIdsToIterate.contains(entry.snapshot().getSnapshotId())) {
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/get/TransportGetSnapshotsAction.java:239]: We can just do this in one go I think:

```java
if (snapshotIdsToIterate.remove(entry.snapshot().getSnapshotId())) {
    snapshotSet.add(new SnapshotInfo(entry));
}
```

?

File: server/src/internalClusterTest/java/org/elasticsearch/snapshots/SnapshotStatusApisIT.java [hunk 2]
```diff
@@ -361,6 +361,49 @@ public void testGetSnapshotsMultipleRepos() throws Exception {
...
     }
 
+    public void testGetSnapshotsWithSnapshotInProgress() throws Exception {
+        createRepository("test-repo", "mock", Settings.builder().put("location", randomRepoPath()).put("block_on_data", true));
+
+        createIndex("test-idx-1");
+        ensureGreen();
+
+        logger.info("--> indexing some data");
+        for (int i = 0; i < 100; i++) {
```
Comment (original-brownbear) [inline server/src/internalClusterTest/java/org/elasticsearch/snapshots/SnapshotStatusApisIT.java:371]: I think you can just use `createIndexWithContent("test-idx-1");` here to get an index with some content. Saving you all the steps including `refresh()`

File: server/src/internalClusterTest/java/org/elasticsearch/snapshots/SnapshotStatusApisIT.java [hunk 3]
```diff
@@ -361,6 +361,49 @@ public void testGetSnapshotsMultipleRepos() throws Exception {
...
+            indexDoc("test-idx-1", Integer.toString(i), "foo", "bar" + i);
+        }
+        refresh();
+
+        logger.info("--> snapshot");
```
Comment (original-brownbear) [inline server/src/internalClusterTest/java/org/elasticsearch/snapshots/SnapshotStatusApisIT.java:376]: No need for this `startFullSnapshot` already logs

File: server/src/internalClusterTest/java/org/elasticsearch/snapshots/SnapshotStatusApisIT.java [hunk 4]
```diff
@@ -361,6 +361,49 @@ public void testGetSnapshotsMultipleRepos() throws Exception {
...
+            .setIgnoreUnavailable(true)
+            .execute().actionGet();
+        List<SnapshotInfo> snapshotInfoList = response1.getSnapshots("test-repo");
+        assertEquals(1, snapshotInfoList.size());
+        assertEquals(SnapshotState.IN_PROGRESS, snapshotInfoList.get(0).state());
+
+        String notExistedSnapshotName = "snapshot_not_exist";
+        GetSnapshotsResponse response2 = client().admin().cluster().prepareGetSnapshots("test-repo").setSnapshots(notExistedSnapshotName)
+            .setIgnoreUnavailable(true)
+            .execute().actionGet();
```
Comment (original-brownbear) [inline server/src/internalClusterTest/java/org/elasticsearch/snapshots/SnapshotStatusApisIT.java:393]: Let's use `.get()` instead to not swallow stack traces (action get unwraps the future's `ExecutionException`).

**Review by original-brownbear — COMMENTED** [review]
Thanks @gaobinlong 

Nice find and great job on the test! Just a few minor points.

File: server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/get/TransportGetSnapshotsAction.java (hunk 1 above)
Comment (gaobinlong) [reply server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/get/TransportGetSnapshotsAction.java]: Yeah, it's better for this situation.

File: server/src/internalClusterTest/java/org/elasticsearch/snapshots/SnapshotStatusApisIT.java (hunk 2 above)
Comment (gaobinlong) [reply server/src/internalClusterTest/java/org/elasticsearch/snapshots/SnapshotStatusApisIT.java]: OK, I've done that.

**Comment by original-brownbear** [comment]
Jenkins test this

**Comment by original-brownbear** [comment]
Thanks @gaobinlong looks great now!

**Review by original-brownbear — APPROVED** [review]

**Pull request merged**